src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+      setValidationMessage(autocompleteRef.current.validationMessage);
+    } else {
+      setValidationMessage(props.validationMessage || '');
+    }
+  }
+
+  function onAutocompleteChange(event) {
+    setExpanded(true);
+    setAutocompleteValue(event.target.value);
+  }
+
+  function onDropdownClose() {
+    setExpanded(false);
+    setAutocompleteValue('');
+  }
+
+  function onOptionRemove(event) {
+    const newValue = value.filter((option, index) => (
+      valueRefs[index].current !== event.target
+    ));
+    setValue(newValue);
+  }
+
+  function onOptionsClear(event) {
+    event.preventDefault();
+
+    if (!props.readOnly) {
+      setValue([]);
+      setExpanded(false);
+      autocompleteRef.current.focus();
+    }
+  }
+
+  function onOptionSelect(event) {
+    const selectedOptionIndex = visibleOptionsRefs.findIndex(optionRef => optionRef === event.target);
+    const selectedOption = visibleOptions[selectedOptionIndex];
+    setExpanded(false);
+    const newValue = [...value, props.optionIDGetter(selectedOption)]
+      .sort((a, b) => props.optionIDGetter(a) - props.optionIDGetter(b));
+    setValue(newValue);
+    setAutocompleteValue('');
+    autocompleteRef.current.focus();
+  }
+
+  function onClick(event) {
+    if (event.defaultPrevented) return;
+    if (props.readOnly) return;
+
+    setExpanded(true);
+  }
+
+  function onKeyDown(event) {
+    switch (event.key) {
+      case 'Escape':
+        setExpanded(false);
+        break;
+      default:
+    }
+  }
+
+  useDropdownClose(wrapperRef, expanded, onDropdownClose);
+
+  useEffect(() => {
+    setValue(props.value);
+  }, [props.value.map(val => JSON.stringify(val)).join(',')]);
+
+  useEffect(() => {
+    props.onChange({ target: selfRef.current });
+  }, [value.map(val => JSON.stringify(val)).join(',')]);
+
+  useImperativeHandle(selfRef, () => ({
+    get dirty() {
+      return props.value.map(val => JSON.stringify(val)).join(',') !== this.value.map(val => JSON.stringify(val)).join(',');
+    },
+    id: props.id,
+    name: props.name,
+    get value() {
+      return value;
+    },
+  }));
+
+  return (
+    <div className={classNames.container} id={props.id} ref={wrapperRef}>
+      <FormControl
+        label={props.label}
+        labelId={ids.label}
+        id={ids.textbox}
+        error={validationMessage}
+        onKeyDown={onKeyDown}
+      >
+        <div role="presentation" className={classNames.formControlChildrenContainer} onClick={onClick}>
+          <TagList
+            className={classNames.tagList}
+            labelledBy={ids.label}
+            refs={valueRefs}
+          >
+            {value.length !== 0 &&
+              props.tagChildren ?
+              props.tagChildren(value, valueRefs, onOptionRemove) :
+              value.map((val, index) => (
+                <Tag
+                  defaultActive={index === 0}
+                  id={`${props.id}-option-${props.optionIDGetter(getOption(val))}`}
```

Comment:
EDIT: This comment might be superflous depending on the discussion of `value` and `label` I pointed out https://github.com/mavenlink/design-system/pull/311#discussion_r634537793

Reply:
Good question! I don't think we should structure all our code based on one-liners from Knuth, but on the other hand, `useMemo` does seem like it might be overkill here... Especially considering it would complicate the component by a not insignificant amount (needing to re-memoize when `value` and `options` change).

Reply:
> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

I think we can just avoid `useMemo` unless we're in a feature space that loads 1000s of objects onto the screen. For a per-component basis, it does not matter.